Zend/zend_enum.c

```diff
@@ -0,0 +1,340 @@
+void zend_enum_register_funcs(zend_class_entry *ce)
+{
+	zend_string *cases_func_name = zend_string_init("cases", strlen("cases"), 1);
+	zend_internal_function *cases_function = malloc(sizeof(zend_internal_function));
+	memset(cases_function, 0, sizeof(zend_internal_function));
+	cases_function->type = ZEND_INTERNAL_FUNCTION;
+	cases_function->module = EG(current_module);
+	cases_function->handler = zend_enum_cases_func;
+	cases_function->function_name = cases_func_name;
+	cases_function->scope = ce;
+	cases_function->fn_flags = ZEND_ACC_PUBLIC|ZEND_ACC_STATIC;
+	cases_function->arg_info = (zend_internal_arg_info *) (arginfo_class_UnitEnum_cases + 1);
+	zend_hash_add_ptr(&ce->function_table, cases_func_name, cases_function);
+
+	if (ce->enum_scalar_type != IS_UNDEF) {
+		zend_string *from_func_name = zend_string_init("from", strlen("from"), 1);
+		zend_internal_function *from_function = malloc(sizeof(zend_internal_function));
+		memset(from_function, 0, sizeof(zend_internal_function));
+		from_function->type = ZEND_INTERNAL_FUNCTION;
+		from_function->module = EG(current_module);
+		from_function->handler = zend_enum_from_func;
+		from_function->function_name = from_func_name;
+		from_function->scope = ce;
+		from_function->fn_flags = ZEND_ACC_PUBLIC|ZEND_ACC_STATIC;
+		from_function->num_args = 1;
+		from_function->required_num_args = 1;
+		from_function->arg_info = (zend_internal_arg_info *) (arginfo_class_ScalarEnum_from + 1);
+		zend_hash_add_ptr(&ce->function_table, from_func_name, from_function);
+
+		zend_internal_function *try_from_function = malloc(sizeof(zend_internal_function));
+		memset(try_from_function, 0, sizeof(zend_internal_function));
+		try_from_function->type = ZEND_INTERNAL_FUNCTION;
+		try_from_function->module = EG(current_module);
+		try_from_function->handler = zend_enum_try_from_func;
+		try_from_function->function_name = zend_string_init("tryFrom", strlen("tryFrom"), 1);
+		try_from_function->scope = ce;
+		try_from_function->fn_flags = ZEND_ACC_PUBLIC|ZEND_ACC_STATIC;
+		try_from_function->num_args = 1;
+		try_from_function->required_num_args = 1;
+		try_from_function->arg_info = (zend_internal_arg_info *) (arginfo_class_ScalarEnum_tryFrom + 1);
+		zend_hash_add_ptr(&ce->function_table, zend_string_init("tryfrom", strlen("tryfrom"), 1), try_from_function);
+	}
+}
+
+void zend_enum_register_props(zend_class_entry *ce)
+{
+	zval name_default_value;
+	ZVAL_UNDEF(&name_default_value);
+	zend_type name_type = ZEND_TYPE_INIT_CODE(IS_STRING, 0, 0);
+	zend_declare_typed_property(ce, ZSTR_KNOWN(ZEND_STR_NAME), &name_default_value, ZEND_ACC_PUBLIC, NULL, name_type);
+
+	if (ce->enum_scalar_type != IS_UNDEF) {
+		zval value_default_value;
+		ZVAL_UNDEF(&value_default_value);
+		zend_type value_type = ZEND_TYPE_INIT_CODE(ce->enum_scalar_type, 0, 0);
+		zend_declare_typed_property(ce, ZSTR_KNOWN(ZEND_STR_VALUE), &value_default_value, ZEND_ACC_PUBLIC, NULL, value_type);
+	}
+}
+
+zval *zend_enum_fetch_case_name(zend_object *zobj)
```

Comment:
nit: Easily fixable: Either make this a static function in zend_enum.h or a ZEND_API method in zend_enum.h and zend_enum.c? (because this is in Zend/, it's ZEND_API instead of PHPAPI)
(I'd prefer the static header function for runtime performance but don't have a strong preference)

```suggestion
ZEND_API zval *zend_enum_fetch_case_name(zend_object *zobj)
```

This is functionality that would be useful to have in serializers such as igbinary/msgpack.

Right now, when I try to call it on a development branch of igbinary on linux to support enum cases, I get ` symbol lookup error: /path/to/modules/igbinary.so: undefined symbol: zend_enum_fetch_case_name` - making this change makes the symbol visible

Additionally, it might be more useful for `zend_enum_fetch_case_name` return the `zend_string *` (possibly add a ZEND_ASSERT on Z_TYPE_P) since I can't imagine name being a non-string?

Reply:
Thoughts?

Reply:
Sorry, I've been kinda taking a break. I was planning on resolving the remaining issues this weekend so we can merge enums next week.

Reply:
These methods are unlikely to change, so I opted for `static zend_always_inline` as you suggested.